storybookjs/storybook
# CSF: Allow overridding globals at the story level

Closes: #23347 #27709 #27524 #27523 

## What I did

- Change the behavior of globals so that `userGlobals` and `storyGlobals` are separate in state.
  - `userGlobals` are globals set via toolbars/addons etc. These are dynamic and change-able.
  - `userGlobals` are initialized using `initialGlobals`.
  - `storyGlobals` are set via a `.globals` property on the `meta` or `story` inside `*.stories.*` files.
  - `storyGlobals` are fixed in place, and cannot be changed.
- `initialGlobals` is introduced as a concept.
  - These can only be set in a `preview.*` file. Addons can set them too.
  - `initialGlobals` are used to determine if a global's value should appear in the URL or not.
- `globals` is still available and results from shallow-merging `userGlobals` and `storyGlobals`.
- the `useGlobals` hook now return a third value: `storyGlobals`.
- decorators now receive an extra context property: `storyGlobals`.

---

- Adddon Viewport is configured slightly differently (see `MIGRATION.md`).
- Addon Backgrounds is configured slightly differently (see `MIGRATION.md`).
- Addon Themes now has an extra parameter `disable`, allowing the disabling of the tool per story.

## Checklist for Contributors

### Testing



#### The changes in this PR are covered in the following automated tests:

#### Manual testing

1. Create & run a sandbox 
2. Visit `localhost:6006/?path=/story/lib-preview-api-globals--overrides` to see basics
3. Visit `localhost:6006/?path=/story/addons-toolbars-globals--override` to see toolbar behaviour
4. Test of the background & viewport addon (they use globals)
5. in the sandbox follow this migration: https://github.com/storybookjs/storybook/blob/6141b5b81a5693cf3eca7f926a6b513bf15b4cd3/MIGRATION.md#L429-L463
6. Now test backgrounds and viewports again
	- take note that the URL should not contain the property/value when the value matches the initialGlobal
	- you should be able to setup initialGlobals in `preview.ts` and change the default value. example:
	  ```ts
	  export const initialGlobals = { viewport: { value: "phone" } };
	  ```
7. Test the inheritance of globals by
	- setup a initialGlobal value
	- set the same global on a meta
	- set the same global on a story
	- expect to see the value set on the story
	- disable the value on the story, and expect the meta's value to be shown.
8. Test the new context values by creating a decorator:
	```ts
	export const decorators = [(StoryFn, { globals, storyGlobals, userGlobals }) => {
		console.log({ globals, storyGlobals, userGlobals });
		return Story();
	}];
    ```
	This should show the correct info when navigating stories, and changing globals


### Documentation



      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release


This pull request has been released as version `0.0.0-pr-26654-sha-49d01553`. Try it out in a new sandbox by running `npx storybook@0.0.0-pr-26654-sha-49d01553 sandbox` or in an existing project with `npx storybook@0.0.0-pr-26654-sha-49d01553 upgrade`.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-26654-sha-49d01553`](https://npmjs.com/package/storybook/v/0.0.0-pr-26654-sha-49d01553) |
| **Triggered by** | @ndelangen |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`tom/23347-story-globals`](https://github.com/storybookjs/storybook/tree/tom/23347-story-globals) |
| **Commit** | [`49d01553`](https://github.com/storybookjs/storybook/commit/49d015536481cc10ea380eb07aaeeaad9751f2fb) |
| **Datetime** | Fri Aug  2 11:46:36 UTC 2024 (`1722599196`) |
| **Workflow run** | [10215034516](https://github.com/storybookjs/storybook/actions/runs/10215034516) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=26654`_
</details>
